### ISSUE TITLE
Exception: BadZipfile: File name in directory "applications\apo_Quote.app" and header "applications/apo_Quote.app" differ.

### ISSUE THREAD

**Issue by Szandor72** [SalesforceFoundation/CumulusCI#530]
Since the latest update, I am experiencing the issue above for various deployments, always getting different errors. (Win10, Powershell)

I assume this to be related to the path differences between Win / Linux/Mac.

**Comment by cdcarter** [comment]
Hey Christian, can you post your full stacktrace?

**Comment by Szandor72** [comment]
Gladly.

```
?[32m2018-01-10 10:51:42:?[0m ?[31mFailing flow due to exception in task?[0m
Traceback (most recent call last):
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\cumulusci\core\flows.py", line 191, in _run_task
    task()
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\cumulusci\core\tasks.py", line 113, in __call__
    self.result = self._run_task()
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\cumulusci\tasks\salesforce.py", line 74, in _run_task
    api = self._get_api()
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\cumulusci\tasks\salesforce.py", line 355, in _get_api
    zipf = self._process_zip_file(zipf)
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\cumulusci\tasks\salesforce.py", line 365, in _process_zip_file
    zipf = self._process_meta_xml(zipf)
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\cumulusci\tasks\salesforce.py", line 403, in _process_meta_xml
    zipf = zip_clean_metaxml(zipf, logger=self.logger)
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\cumulusci\utils.py", line 256, in zip_clean_metaxml
    content = zip_src.read(name)
  File "c:\python27\Lib\zipfile.py", line 938, in read
    return self.open(name, "r", pwd).read()
  File "c:\python27\Lib\zipfile.py", line 983, in open
    zinfo.orig_filename, fname)
BadZipfile: File name in directory "applications\apo_Quote.app" and header "applications/apo_Quote.app" differ.
Exception IOError: (0, 'Error') in <bound method ZipFile.__del__ of <zipfile.ZipFile object at 0x0000000005520940>> ignored
Traceback (most recent call last):
  File "c:\python27\Lib\runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "c:\python27\Lib\runpy.py", line 72, in _run_code
    exec code in run_globals
  File "C:\Users\szand\.virtualenvs\cumulus\Scripts\cci.exe\__main__.py", line 9, in <module>
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\click\core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\click\core.py", line 697, in main
    rv = self.invoke(ctx)
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\click\core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\click\core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\click\core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\click\core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\cumulusci\cli\cli.py", line 230, in new_func
    return ctx.invoke(f, obj, *args[1:], **kwargs)
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\click\core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\cumulusci\cli\cli.py", line 1091, in flow_run
    handle_exception_debug(config, debug, e, no_prompt=no_prompt)
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\cumulusci\cli\cli.py", line 1067, in flow_run
    flow()
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\cumulusci\core\flows.py", line 129, in __call__
    self._run_task(flow_task_config)
  File "c:\users\szand\.virtualenvs\cumulus\lib\site-packages\cumulusci\core\flows.py", line 200, in _run_task
    raise e
zipfile.BadZipfile: File name in directory "applications\apo_Quote.app" and header "applications/apo_Quote.app" differ.
```

**Comment by dmitry-sforce** [comment]
Hey guys, having the same issue with the latest update downloaded yesterday. OS win10, fails on "deploy" task. Other tasks, i.e. running tests, work fine, not sure about all of them though. The previous version does not have this error.

**Comment by jlantz** [comment]
As a workaround for now, you can set the `clean_meta_xml` option on the deploy task to False to disable the meta.xml rewriting logic that was introduced in that release.  We're working on another release that will fix the underlying issue.

**lurch: add

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x80000009KL8SAAW" target="_blank">W-022286</a>

**Comment by Szandor72** [comment]
Hey @jlantz 

any thoughts on this? Can I support you?

Id run stuff in my win  bash but I am having issues with the current ubuntu +  openssl + python 2.7

Id love to see CCI running on win again.

**Comment by e02d96ec16** [comment]
##### TL;DR
My workaround:
```diff
--- <your python location>\Lib\zipfile.py
+++ <your python location>\Lib\zipfile.py
@@ -974,7 +974,7 @@
             if fheader[_FH_EXTRA_FIELD_LENGTH]:
                 zef_file.read(fheader[_FH_EXTRA_FIELD_LENGTH])
 
-            if fname != zinfo.orig_filename:
+            if fname != zinfo.orig_filename and os.path.normpath(fname) != zinfo.orig_filename:
                 raise BadZipfile, \
                         'File name in directory "%s" and header "%s" differ.' % (
                             zinfo.orig_filename, fname)
```

##### Story

I ran into this a while ago. My investigation led me to believe this is happening due to [a (open) bug in python](https://bugs.python.org/issue6839).

The [Zip file specification](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT) actually demands forward slashes only:

> 4.4.17 file name: (Variable)
>
>       4.4.17.1 The name of the file, with optional relative path.
>       The path stored MUST not contain a drive or
>       device letter, or a leading slash.  All slashes
>       MUST be forward slashes '/' as opposed to
>       backwards slashes '\' for compatibility with Amiga
>       and UNIX file systems etc.  If input came from standard
>       input, there is no file name field.

From my meagre understanding of CumulusCI, when deploying, CCI will actually create a [deploy bundle](https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/file_based_zip_file.htm) which is essentially a zip file.
Then CCI will want to perform certain modification on said bundle e.g. remove/replace namespace placeholders, remove package versions.
The way it does this is by iterating over the already gathered file paths from back when it generated the zip (i.e. value reuse; it already knows what the files are, it doesn't need to find out again).
This iteration in turn fails because the files-on-disk path doesn't look like the file-in-zip path.

Btw, this zip never rests on the disk: it's all (generation; manipulation) happening in-memory. I don't want to admit how long it took me to get a hold of one.

##### Conclusion

Technically this could be classified as a "known issue" due to the python bug.
But it could also be argued it's a bug in CCI: incorrectly handling of zip files.
I'm wondering if switching to iterating over the zip-file files instead of the on-disk ones wouldn't fix this.

As for the python bug, although the fix appears straight-forward, I have no idea what other bits in terms of tests (or builds) it would require. If anyone wishes to submit it, feel free.

**Comment by cdcarter** [comment]
@e02d96ec16 thank you SO much for your investigations! your description of what CCI does is correct, we manage all the zipfile madness in memory and use it as a way to do transforms on the metadata.

@cpolcyn will be back from PTO next week and can take a look at your idea/suggestions. Getting this bug fixed is a big priority for April. thax

**Comment by e02d96ec16** [comment]
✅ **Confirmed**

Verified using `cci task run deploy -o namespace_strip True` (the option is just for good measure): failed in v90, worked in v91.

🙇 Thanks for your work!

**Comment by cdcarter** [comment]
@sam-knox - you should be good to update cci and try moving forward with the tutorial, when you have time :)

**Comment by sam-knox** [comment]
@cdcarter - I think I'm good to go now. I was able to run cci flow run dev_org which I don't think I could before.
